lib/Parse/ParseExpr.cpp

```diff
@@ -2958,8 +2958,8 @@ ExprResult Parser::ParseBoundsExpression() {
   return Result;
 }
 
-bool Parser::ParseRelativeBoundsClause(ExprResult &Expr) {
-  bool IsError = false;
+RelativeBoundsClause *Parser::ParseRelativeBoundsClause(ExprResult &Expr,
```

Comment:
I think this interface is too complicated now.   The function is doing too many things. 

I  suggest that you remove ExprResult as an argument and replace it with a Boolean flag that indicates whether this method should print an error message.   You can move the code that sets the bounds property of the ExprResult to the caller.

Then the function becomes easier to explain:
- It tries to parse a relative bounds expression.  It returns the relative bounds expression if successful and null if there was an error.
- If it cannot parse a relative bounds expression, it optionally prints an error message that it expected a relative bounds expression, if the new flag says it should.